faculty/clients/serveragent.py

```diff
@@ -0,0 +1,344 @@
+# Copyright 2018-2021 Faculty Science Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Interact with server agent.
+"""
+from contextlib import contextmanager
+import json
+from enum import Enum
+
+from attr import attrs, attrib
+from marshmallow import fields, post_load
+from marshmallow_enum import EnumField
+
+from faculty.clients.base import BaseSchema, BaseClient
+
+SERVER_RESOURCES_EVENT = "@SSE/SERVER_RESOURCES_UPDATED"
+
+class ExecutionStatus(Enum):
+    """The status of an environment execution."""
+    NOT_STARTED = "NOT_STARTED"
+    STARTED = "STARTED"
+    SUCCESS = "SUCCESS"
+    FAILURE = "FAILURE"
+
+class EnvironmentExecutionStepStatus(Enum):
+    """The status of an environment execution step."""
+    QUEUED = "QUEUED"
+    CANCELLED = "CANCELLED"
+    STARTED = "STARTED"
+    SUCCESS = "SUCCESS"
+    FAILURE = "FAILURE"
+
+@attrs
+class Execution(object):
+    """Server environment execution.
+    
+    Parameters
+    ----------
+    id : uuid.UUID
+        The ID of the execution.
+    status : enum.Enum
+        The status of the execution.
+    environments : List[EnvironmentExecution]
+        list of EnvironmentExecution objects for each environment applied.
+    started_at : Optional[datetime]
+        startime of the execution.
+    finished_at : Optional[datetime]
+        endtime of the execution.
+    """
+
+    id = attrib()
+    status = attrib()
+    environments = attrib()
+    started_at = attrib()
+    finished_at = attrib()
+
+
+@attrs
+class EnvironmentExecution(object):
+    """An environment executed on a server.
+    
+    Parameters
+    ----------
+    id : uuid.UUID
+        The ID of the environment.
+    steps : List[EnvironmentExecutionStep]
+        List of EnvironmentExecutionStep objects.
+    """
+
+    id = attrib()
+    steps = attrib()
+
+
+@attrs
+class EnvironmentExecutionStep(object):
+    """A single environment execution step on a server.
+    
+    Parameters
+    ----------
+    id : uuid.UUID
+        The ID of the environment step.
+    command : List[str]
+        The command executed by the step.
+    status : enum.Enum
+        The status of the step execution.
+    started_at : Optional[datetime]
+        The start time of the execution step.
+    finished_at : Optional[datetime]
+        The finish time of the execution step.
```

Comment:
Minor edit for consistency with other parameters of this class:

```suggestion
        The finish time of the environment execution step.
```